### ISSUE TITLE
Sticky port - stable upon restart per repository

### ISSUE THREAD

**Issue by mihailik** [parcel-bundler/parcel#1628]
# 💬 RFC

Running multiple parcel projects side by side in `watch` mode is automatically possible — but ports are randomized.

Would be great if the chosen port was sticky to the project:
* restarting same project again — hosts on the same port
* but still possible running multiple parcel builds side by side

## 🔦 Suggestion

That could be done in two different ways:
1. derive port from hash of directory path
2. when random port picked, preserve it in `.cache` directory

I prefer the first option: it's simpler and less jittery.

Mind you, the famous `1234` default would be ditched if that option is used.

But in spirit of simplicity and 'works out of the box' philosophy of `parcel` that definitely boosts the value proposition!

## Caveat: twice from same directory

When `parcel` is fired from the same repository twice, currently that runs two instances in parallel. Even it doesn't mess up `.cache`, it still can confuse developers.

Suggestion is to send a special termination HTTP request from newer instance of `parcel` to the old. They share the port, so the URL is known. Upon receiving a termination request, `parcel` DEV HTTP server will shut down the instance, bequeathing the port to the new instance.

The DEV HTTP startup logic would go:
* determine the sticky port number
* try to listen - if succeeds, DONE
* if listen failed, send HTTP POST to the same port/host with custom 'termination' request payload
* keep trying to listen to the sticky port — until the original `parcel` instance exists and yields the port

## 💻 Examples

I clone my project and work on it for weeks. It sits in the same directory, and its path-hash-derived port allocated is 51410.

While that work is ongoing, I clone the project for a urgent production patch in another directory. Parcel allocates path-hash-derived port 58013. Good, no confusion!

Without 'sticky' ports it is too easy to get confused between browser tab pointing to one port and another. Which one is my main branch of work, which one is the urgent prod patch?

And if I need to restart `parcel`, the old port in that browser tab is no longer valid (as each new parcel instance start randomizes the port).

**Comment by DeMoorJasper** [comment]
I think this isn't something that should be in parcel by default. I think the `--port` flag fits perfect with this use-case

**Comment by mihailik** [comment]
Could you please elaborate?

Port flag requires user intervention to start with. It's actually too sticky: if you put `--port` option in `package.json`'s script section, it will be the same port whenever you clone that repo.

Hence it actually doesn't fit the master/patch branch case mentioned.

There's already a feature to pick randomized port number, my suggestion is to improve upon that feature.

**Comment by DeMoorJasper** [comment]
I didn't mean in the package.json, i meant in the command

for example:
```Json
{
  "scripts": {
    "start": "parcel index.html"
  }
}
```

```bash
yarn start --port 8800
```
```bash
yarn start --port 8801
```
```bash
yarn start --port 8802
```

**Comment by mihailik** [comment]
Got it, but doesn't look too smooth!

The biggest value proposition of `parcel` is taking away the hassle. Implementing sticky ports in my head, instead of offloading that mundane job to machine is such a hassle.

Hope that can be noted under 'low priority improvements' or something?
